### ISSUE TITLE
Variance and Nested Protocols

### ISSUE THREAD

**Issue by colehaus** [microsoft/pyright#3207]
When using a certain set of nested protocols, pyright declares a seemingly correct implementation of that protocol incorrect due to variance issues. In particular:
```py
from dataclasses import dataclass
from typing import Protocol, TypeVar

X = TypeVar("X", covariant=True)
Y = TypeVar("Y", contravariant=True)
Z = TypeVar("Z")

class Fn(Protocol[Y]):
    def __call__(self, y: Y) -> None: ...
class FnHandler(Protocol[X]):
    def __call__(self, x: Fn[X]) -> None: ...
def handler(x: Fn[int]) -> None: ...

class ModuleSpec(Protocol[Z]):
    x: FnHandler[Z]
@dataclass
class Module:
    x = handler

h: FnHandler[int] = Module().x
m: ModuleSpec[int] = Module() # error here
```
produces an error on the final line:
```
 error: Expression of type "Module" cannot be assigned to declared type "ModuleSpec[int]"
    "Module" is incompatible with protocol "ModuleSpec[int]"
      "x" is an incompatible type
        Type "(x: Fn[int]) -> None" cannot be assigned to type "(x: Fn[Z@ModuleSpec]) -> None"
          Parameter 1: type "Fn[Z@ModuleSpec]" cannot be assigned to type "Fn[int]"
            TypeVar "Y@Fn" is contravariant (reportGeneralTypeIssues)
```
This seems pretty surprising given that the line immediately above it type checks fine and all `Module` is doing is passing an invariant `TypeVar` down to the single member. A further curiosity is that the following code does type check wherein we ignore the demand to make our `TypeVar` contravariant:

```py
from dataclasses import dataclass
from typing import Protocol, TypeVar

Z = TypeVar("Z")

class Fn2(Protocol[Z]): # type: ignore # Suppress complaint that Z should be contravariant
    def __call__(self, y: Z) -> None: ...
class FnHandler2(Protocol[Z]):
    def __call__(self, x: Fn2[Z]) -> None: ...
def handler2(x: Fn[int]) -> None: ...

class ModuleSpec2(Protocol[Z]):
    x: FnHandler2[Z]
@dataclass
class Module2:
    x = handler

h2: FnHandler2[int] = Module2().x
m2: ModuleSpec2[int] = Module2()
```
My understanding is that contravariant and covariant types should always be more permissive than invariant, but in this case the invariant `TypeVar` permits the program to check while the variant version did not type check.

But maybe I'm misunderstanding something and this is all operating as expected.

I'm using pyright 1.1.230 on the command line.

**Comment by erictraut** [comment]
Pyright is correct in flagging this as an error, but the error message is less than ideal and is obfuscating the real issue.

The problem is that when you assign `x = handler` in the `Module` class, you're creating an instance method. The parameter `x` in the `handler` function effectively becomes a `self` at runtime.

You can see this by running the following, which prints out `Method()`.
```python
from dataclasses import dataclass

def handler(x) -> None:
    print(x)

@dataclass
class Module:
    x = handler

m = Module()
m.x()
```

The type violation occurs when pyright attempts to bind the function `Module.x` to its containing class `Method`, effectively trying to assign `ModuleSpec` to the "self" parameter of `handler` which is annotated as `Fn[int]`.

The correct fix is to mark `handler` as a static method. If you do this, the type error goes away and the code runs without an exception at runtime.
```python
@dataclass
class Module:
    x = staticmethod(handler)
```

Incidentally, mypy doesn't handle this code correctly. It emits a false positive error because of a known bug.

**Comment by colehaus** [comment]
Thanks for the reply!

I tried to simplify my actual use case and introduced the error you pointed out during that process. My actual use case involves type checking a module against a protocol:

`scratch2.py`:
```python
from typing import Protocol, TypeVar

Y = TypeVar("Y", contravariant=True)
class Fn(Protocol[Y]):
    def __call__(self, y: Y) -> None: ...

def x(x: Fn[int]) -> None:
    print(x)
```
`scratch.py`:
```py
from typing import Protocol, TypeVar

import scratch2
from scratch2 import Fn

X = TypeVar("X", covariant=True)
Z = TypeVar("Z")

class FnHandler(Protocol[X]):
    def __call__(self, x: Fn[X]) -> None: ...
class ModuleSpec(Protocol[Z]):
    x: FnHandler[Z]

z: ModuleSpec[int] = scratch2 // error here

z.x(lambda y: None)
```

I get an error similar to the last one:
```
error: Expression of type "Module("scratch2")" cannot be assigned to declared type "ModuleSpec[int]"
    "x" is an incompatible type
      Type "(x: Fn[int]) -> None" cannot be assigned to type "(x: Fn[Z@ModuleSpec]) -> None"
        Parameter 1: type "Fn[Z@ModuleSpec]" cannot be assigned to type "Fn[int]"
          TypeVar "Y@Fn" is contravariant
            Type "int" cannot be assigned to type "Z@ModuleSpec" (reportGeneralTypeIssues)
```
I don't believe there are any `self` issues here and indeed when I run this code, I get the expected `print` output about a lambda function. And again, marking all the `TypeVar`s as invariant here (and ignoring the prompt to make `Y` contravariant) eliminates all errors from pyright.

Am I making another silly error here or is this an actual issue with pyright?

Thanks again.

**Comment by erictraut** [comment]
That's a real bug. The code path for matching modules against protocols was not properly handling the case where a generic protocol was explicitly specialized, as in your example. This will be fixed in the next release.

**Comment by colehaus** [comment]
Thanks!

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.231, which I just published. It will also be included in the next release of pylance.
